### PR TITLE
Issue 5330: SLTS - Fix NullPointerException in ReadIndexCache::evictChunks and avoid re-entry in eviction code.

### LINKED ISSUE #5347: (SLTS) NullPointerException in ReadIndexCache::evictChunks
**Problem description**
NullPointerException is observed ReadIndexCache::evictChunks

**Problem location**
ReadIndexCache::evictChunks

**Suggestions for an improvement**
Check for null value

**Logs**
```
2020-11-04 05:01:07,734 79467372 [core-5] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[6]: Iteration[125261].Error.
java.lang.NullPointerException: null
        at io.pravega.segmentstore.storage.chunklayer.ReadIndexCache.evictChunks(ReadIndexCache.java:337)
        at io.pravega.segmentstore.storage.chunklayer.ReadIndexCache.evictChunks(ReadIndexCache.java:292)
        at io.pravega.segmentstore.storage.chunklayer.ReadIndexCache.addIndexEntry(ReadIndexCache.java:158)
        at io.pravega.segmentstore.storage.chunklayer.ReadOperation.lambda$findChunkForOffset$11(ReadOperation.java:205)
        at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source)
        at java.base/java.util.concurrent.CompletableFuture$Completion.run(Unknown Source)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
```

### PR DESCRIPTION
**Pull request by sachin-j-joshi**
Issue 5330: SLTS - Fix NullPointerException in ReadIndexCache::evictChunks and avoid re-entry in eviction code.
Signed-off-by: Sachin Joshi <sachin.joshi@emc.com>

**Change log description**  
Fix NullPointerException in ReadIndexCache::evictChunks in case of possible race condition in concurrent evictions/additions.

**Purpose of the change**  
Fixes #5347 

**What the code does**  
Checks null.
Allows only one eviction in flight per segment.
Minor code cleanup.

**How to verify it**  
All existing tests should continue to pass.

### CODE REVIEW

**Comment by sachin-j-joshi** [comment]
It is hard to recreate exact race condition in unit tests that lead to NPE.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadIndexCache.java [hunk 1]
```diff
@@ -362,11 +373,14 @@ private void evictChunks(SegmentReadIndex segmentReadIndex, long toRemoveCount)
      * The underlying data structure uses ConcurrentSkipListMap.
      * The generation is used to cleanup unused entries.
      */
-    @Builder
-    @Data
+    @RequiredArgsConstructor
```
Comment (andreipaduroiu) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadIndexCache.java:376]: `@Data` is the same as `@RequiredArgsConstructor` + `@Getter`. You could have left that and save the extra Getters below.

Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadIndexCache.java:375]: ok. added back @Data

**Review by RaulGracia — CHANGES_REQUESTED** [review]
While the change is small, try to add some tests to validate the fix (there are tests in the project that do an effort to check concurrency issues, so maybe we need to try the same here).

**Comment by sachin-j-joshi** [comment]
Closing this as it should be superseded by https://github.com/pravega/pravega/pull/5363

**Pull request closed without merging**